packages/web3-eth/src/utils/get_revert_reason.ts

```
@@ -0,0 +1,84 @@
+ * @returns `undefined` if no revert reason was given, a revert reason object, a revert reason string, or an `unknown` error
+ */
+export async function getRevertReason<
+	ReturnFormat extends DataFormat = typeof DEFAULT_RETURN_FORMAT,
+>(
+	web3Context: Web3Context<EthExecutionAPI>,
+	transaction: TransactionCall,
+	contractAbi?: ContractAbi,
+	returnFormat: ReturnFormat = DEFAULT_RETURN_FORMAT as ReturnFormat,
+): Promise<undefined | RevertReason | RevertReasonWithCustomError | string | unknown> {
+	try {
+		await call(web3Context, transaction, web3Context.defaultBlock, returnFormat);
+		return undefined;
+	} catch (error) {
+		if (
+			error instanceof ContractExecutionError &&
+			error.innerError instanceof Eip838ExecutionError
+		) {
+			if (contractAbi !== undefined) {
+				const errorsAbi = contractAbi.filter(abi =>
+					isAbiErrorFragment(abi),
+				) as unknown as AbiErrorFragment[];
+				decodeContractErrorData(errorsAbi, error.innerError);
+
+				return {
+					reason: error.innerError.message,
+					signature: error.innerError.data?.slice(0, 10),
+					data: error.innerError.data?.substring(10),
+					customErrorName: error.innerError.errorName,
+					customErrorDecodedSignature: error.innerError.errorSignature,
+					customErrorArguments: error.innerError.errorArgs,
+				} as RevertReasonWithCustomError;
+			}
+
+			return {
+				reason: error.innerError.message,
+				signature: error.innerError.data?.slice(0, 10),
+				data: error.innerError.data?.substring(10),
+			} as RevertReason;
+		}
+
+		if (
+			error instanceof InvalidResponseError &&
+			!Array.isArray(error.innerError) &&
+			error.innerError !== undefined
+		) {
+			return error.innerError.message;
+		}
+
+		return error instanceof Error ? error.message : error;
```

In case of generic error ( that is not due to failed tx reverted ) , we should throw it and remove `string` from return type`Promise<undefined | RevertReason | RevertReasonWithCustomError | string | unknown>`. 

Because for `getRevertReason` function call with eth_call rpc in case of generic errors: 
1. neither tx eth_call was reverted with reason.
2. nor it succeeded with out any revert reason, 
but request for looking any of above answers failed, that user should catch. It will also make return types homogeneous.

---

Updated via [this commit](https://github.com/web3/web3.js/pull/5844/commits/28a3d848830a0e17aa706b5092c1e917e1b87f02), however `string` is still a possible return type because of

```typescript
if (
	error instanceof InvalidResponseError &&
	!Array.isArray(error.innerError) &&
	error.innerError !== undefined
) {
	return error.innerError.message;
}
```

---

if InvalidResponseError is also not categorizable under 1 and 2, shouldn't it also be thrown?

---

@jdevcs The code snippet above is for errors such as

```typescript
InvalidResponseError: Returned error: err: intrinsic gas too low: have 0, want 21544 (supplied gas 0)
          at Web3RequestManager._processJsonRpcResponse (/home/anon/Public/code/ChainSafe/git-repos/web3.js/packages/web3-core/src/web3_request_manager.ts:372:11)
          at Web3RequestManager.<anonymous> (/home/anon/Public/code/ChainSafe/git-repos/web3.js/packages/web3-core/src/web3_request_manager.ts:216:16)
          at Generator.next (<anonymous>)
          at fulfilled (/home/anon/Public/code/ChainSafe/git-repos/web3.js/packages/web3-core/lib/web3_request_manager.js:5:58)
          at processTicksAndRejections (node:internal/process/task_queues:95:5) {
        innerError: {
          code: -32000,
          message: 'err: intrinsic gas too low: have 0, want 21544 (supplied gas 0)'
        },
        code: 101,
        data: undefined
      }
```

---

also it can be generic error like : https://github.com/web3/web3.js/blob/b948ea4f3c6d48509a7aa726a2d051a8c83ebc96/packages/web3-utils/src/chunk_response_parser.ts#L67.  ( I think in that case at least , it should be modified in chunkresponseparser to be some different error type and code ) and we can keep returning InvalidResponseError as it is here.

---

@jdevcs That's a good point. I think the error in the chunk parser should remain `InvalidResponseError`, but errors like the gas too low from above should be renamed to something like `RpcError`

If you agree, I can create an issue for it, merge this PR, then make the error name change?

---

sure. Good work @spacesailor24